### ISSUE TITLE
Interfaces extending class with private props

### ISSUE THREAD

**Issue by vidartf** [jupyterlab/jupyterlab#6333]
**Describe the bug**
When working on #6302, I tried to make an extension that would replace the document manager by providing the `IDocumentManager` token. So I make a new class that implements the interface, but immediately get hit by the error:

```ts
error TS2420: Class 'MyDocumentManager' incorrectly implements interface 'IDocumentManager'.
  Types have separate declarations of a private property '_findContext'.
```

It is not an alternative to not declare it, as it will then complain about the missing private property (!).

**To Reproduce**

Any interface declaration extending a class definition with private properties will throw this error, e.g.:
```ts
interface ISecrectInterface {};

class A {
  private _method(): ISecrectInterface {
    return {};
  }
}

interface IA extends A {};

class B implements IA {
  
}

```

**Expected behavior**
Exported interfaces, and especially token interfaces, should not include any private property declaration.

**Comment by vidartf** [comment]
Possible solutions:
- Use some TS magic to ensure interface only extends *public* properties of class.
- Maintain the interface declaration separately from the class definitions, i.e. stop using the `export interface IInterface extends CClass`.

**Comment by vidartf** [comment]
Note: Could this be a bug in TS?

**Comment by jasongrout** [comment]
I totally agree that tokens should be exporting explicit interfaces, not interfaces derived from implementation classes.

**Comment by jasongrout** [comment]
Putting as 1.0 since a stable extension API is one of our goals for 1.0.

**Comment by vidartf** [comment]
TS magic: 

```ts
type PublicShape<T> = Pick<T, keyof T>;

export interface IInterface extends PublicShape<CClass> {};
```

**Comment by jasongrout** [comment]
`Pick` only takes public attributes?

**Comment by vidartf** [comment]
According to the issue here: https://github.com/microsoft/TypeScript/issues/31299

However, I am not certain if this is documented, or just "current behavior".

**Comment by jasongrout** [comment]
Ah, `keyof` currently skips private members. In this usecase, protected members probably also should be skipped?

I think it is much clearer if pure interfaces are the explicit public contract, and implementation classes always implement the token interfaces.

**Comment by vidartf** [comment]
Note: It seems super weird to me the interface picks up the private properties of the class, since declaring a private property on an interface gives a compile error. This is why I'm wondering if this is a TS bug.

**Comment by jasongrout** [comment]
Can you make a list of the token interfaces derived from implementation classes? I don't think this is a widespread problem - I think most places tokens are typed with pure interfaces.

**Comment by jasongrout** [comment]
> This is why I'm wondering if this is a TS bug.

Sounds like a good issue to file with them.

**Comment by vidartf** [comment]
Xref: https://github.com/microsoft/TypeScript/issues/471

**Comment by jasongrout** [comment]
https://github.com/microsoft/TypeScript/issues/18499 has some relevant discussion as well.

**Comment by vidartf** [comment]
I think this is simply a quirk we will have to work around, given that `publicShape` is the conclusion in the xref issue I linked above. TL;DR: The issue is because the `implements` keywords can also be used directly between classes, and going via an interface seems to resolve down to this. I'm not going to bother to take this upstream at this point.

Note: Some places in the Lab code, this behavior is probably what we want. E.g.
```ts
export interface IShell extends Widget { ... }
```

Here, I think we definitely want to ensure that a replacement of `IShell` is a `Widget` with a solid identity to the known `Widget`.

**Comment by vidartf** [comment]
My conclusion here: We should only use the `interface I extends C {}` if we want to ensure that any implementer of `I` subclasses `C`.

**Comment by vidartf** [comment]
Current interfaces (mainly tokens) that should have a closer look:
- `ILabShell` (token)
- `IThemeManager` (token)
- `ISettingsRegistry` (token)
- `docmanager/manager: Private.IContext` (private type, so unsure if it matters)
- `IDocumentManager` (token)
- `IRenderMimeRegistry` (token)
- `Builder.IManager`
- `NbConvert.IManager`
- `Setting.IManager`
- `Workspace.IManager`

I don't know how the `IManager` pattern is generally used, so it might fall into either category (subclassing is required or not).

Recommendation: Extract the interface of all tokens, and revert the logic (interface declaration is explicit, and class `implements` that). This will add some friction when changing the interface of a Token, which is probably a Good Thing.

**Comment by blink1073** [comment]
IMO, I don't think we ever should explicitly expect a subclass at the token level, only potentially within our own class hierarchies.

**Comment by vidartf** [comment]
Current tokes that `extends Widget` (not guaranteed to be exhaustive):
- `INotebookTools`
- `IShell`
- `ILabShell` (extends `LabShell` which extends `Widget`)

**Comment by vidartf** [comment]
Note: I can do the legwork here, but I need input on:
- 👍 / 👎 on proposed solution for (non-Widget) tokens:
> Extract the interface of all tokens, and revert the logic (interface declaration is explicit, and class implements that). This will add some friction when changing the interface of a Token, which is probably a Good Thing.
- What to do with the tokens that `extends Widget`.

**Comment by afshin** [comment]
`ILabShell` is very specifically meant to be *the* Lab application shell, so the fact that it `extends LabShell` seems okay to me. It's not the "generic Jupyter front-end" shell. What do you think?

**Comment by vidartf** [comment]
@blink1073 For the three Widget/token mixes outlined above, would you say that it makes sense to have the requirement that the provided object subclasses `Widget`? I'm guessing it would be a bit of work to have the shell not be a widget, and it seems deeply ingrained for `INotebookTools` as well.

**Comment by jasongrout** [comment]
I think it is okay to assume `extends Widget` should be a subclass. I think there is an explicit decision in phosphor design that you extend there by subclassing Widget, not implementing a Widget interface.

**Comment by blink1073** [comment]
:+1: for what @jasongrout said

**Comment by vidartf** [comment]
I ended up adding the fix for this together with that of #6302 in #6334, as they are somewhat related.
